lightning/src/ln/channel.rs

```diff
@@ -3878,7 +3880,8 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				htlcs.push(htlc);
 			}
 
-			let res = self.local_keys.sign_remote_commitment(feerate_per_kw, &remote_commitment_tx.0, &remote_keys, &htlcs, &self.secp_ctx)
+			let pre_remote_keys = PreCalculatedTxCreationKeys::new(remote_keys.clone());
```

Comment:
Nit, you can avoid the clone() here (which probably allows the compiler to avoid copying the full memory pretty trivially).
```
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 1ae894f7..16345265 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -3881,9 +3881,10 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
                        }
 
-                       let pre_remote_keys = PreCalculatedTxCreationKeys::new(remote_keys.clone());
+                       let pre_remote_keys = PreCalculatedTxCreationKeys::new(remote_keys);
                        let res = self.local_keys.sign_remote_commitment(feerate_per_kw, &remote_commitment_tx.0, &pre_remote_keys, &htlcs, &self.secp_ctx)
                                .map_err(|_| ChannelError::Close("Failed to get signatures for new commitment_signed".to_owned()))?;
                        signature = res.0;
                        htlc_signatures = res.1;
+                       let remote_keys = pre_remote_keys.trust_key_derivation();
 
                        log_trace!(logger, "Signed remote commitment tx {} with redeemscript {} -> {}",
```

Reply:
If it's moved here, it can't be borrowed below.

Reply:
ah, yes.